### PR TITLE
Make output cert readable to other containers running on kube.

### LINKED ISSUE #28256: Add integration test for secured prometheus scraping
Background: https://github.com/istio/istio/issues/28050.

As we are adding guide about how to use sidecar provision certs for prometheus secured scraping, we need to add an integration test or convert the existing test to cover that.

### PR DESCRIPTION
**Pull request by bianpengyuan**
Fix https://github.com/istio/istio/issues/29919
Fix #28256

### CODE REVIEW

**Review by howardjohn — COMMENTED** [review]
looks good

File: security/pkg/nodeagent/util/util.go [hunk 1]
```diff
@@ -25,8 +25,12 @@ import (
 	"go.opencensus.io/stats/view"
 
 	"istio.io/istio/pkg/file"
+	"istio.io/pkg/env"
 )
 
+var k8sInCluster = env.RegisterStringVar("KUBERNETES_SERVICE_HOST", "",
+	"Kuberenetes service host, set automatically when running in-cluster")
```
Comment (lei-tang) [inline security/pkg/nodeagent/util/util.go:32]: Can you point to the code that sets KUBERNETES_SERVICE_HOST in k8s environment for Istio Agent?

Comment (howardjohn) [reply security/pkg/nodeagent/util/util.go:32]: kubernetes sets it in all pods

File: security/pkg/nodeagent/util/util.go [hunk 2]
```diff
@@ -70,23 +74,29 @@ func OutputKeyCertToDir(dir string, privateKey, certChain, rootCert []byte) erro
 	if len(dir) == 0 {
 		return nil
 	}
+
+	certFileMode := os.FileMode(0600)
+	if k8sInCluster.Get() != "" {
+		// If this is running on k8s, give more permission to the file certs.
```
Comment (howardjohn) [inline security/pkg/nodeagent/util/util.go:80]: can you update the comment to explain *why* we do this?

**Review by howardjohn — APPROVED** [review]

File: security/pkg/nodeagent/util/util.go (hunk 2 above)
Comment (bianpengyuan) [reply security/pkg/nodeagent/util/util.go:80]: Done.

**Comment by bianpengyuan** [comment]
@istio/wg-policies-and-telemetry-maintainers ptal thanks!

**Comment by bianpengyuan** [comment]
Ping @istio/wg-policies-and-telemetry-maintainers  ptal thanks!

**Comment by bianpengyuan** [comment]
@mandarjog @douglas-reid ptal. this is a upgrade breaking change and would be great to get in asap...

**Review by nrjpoddar — COMMENTED** [review]
@bianpengyuan do our integration tests not use this capability? I’m trying to understand why this wasn’t caught in our testing before the release.

**Comment by bianpengyuan** [comment]
> do our integration tests not use this capability? I’m trying to understand why this wasn’t caught in our testing before the release.

It does not. This PR adds an integration test for this.

**Comment by nrjpoddar** [comment]
I see. So we changed the permissions if output certs in 1.9 without understanding the impact? What was the motivation or change?

**Comment by bianpengyuan** [comment]
AFAIK, OUTPUT_CERTS is also used at VM, the stricter permission is needed there to make sure only Envoy can access the certs, but not other process running on the VM. @liminw would have more detail on this.

**Review by nrjpoddar — APPROVED** [review]
Thanks! This LGTM.

**Comment by bianpengyuan** [comment]
@howardjohn https://github.com/istio/istio/pull/31374 makes prom secured scraping not working any more. Though the fix looks reasonable, I don't know if there is a way to distinguish prom scrape with Istio certs vs passthrough standard TLS.

**Comment by bianpengyuan** [comment]
One hacky but feasible way is to add a special filter chain match for a hard coded SNI for prom secured scraping. Prom tls config can set server_name.

**Comment by howardjohn** [comment]
If I understand correctly, if that change broke prom scraping then it never
really worked fully? Because that makes the passthrough logic the same as
Service port logic, so that means it should only work if they have didn't
have a Service?

**Comment by bianpengyuan** [comment]
From description PR, IIUC it worked before because we matched standard alpn `http/1.1` for istio downstream tls context before?

```diff
+{"prefixRanges":[{"addressPrefix":"0.0.0.0","prefixLen":0}],"transportProtocol":"tls"}
+{"prefixRanges":[{"addressPrefix":"0.0.0.0","prefixLen":0}],"transportProtocol":"tls","applicationProtocols":["istio-http/1.0","istio-http/1.1","istio-h2"]}
-{"prefixRanges":[{"addressPrefix":"0.0.0.0","prefixLen":0}],"transportProtocol":"tls","applicationProtocols":["http/1.0","http/1.1","h2c","istio-http/1.0","istio-http/1.1","istio-h2"]}
```

**Comment by howardjohn** [comment]
Yeah but only on the passthrough, the PR makes passthrough match the
service port config. So I think that if they have the scraping port exposed
in a Service it already wouldn't work?

btw it's only broken in permissive, strict mode should work

**Comment by bianpengyuan** [comment]
> btw it's only broken in permissive, strict mode should work

Yeah good point. I think we'd need to update doc so that during permissive transition period, user can use metrics merging, and switch to tls scraping after turning on strict mode.

**Pull request merged**